### ISSUE TITLE
Updates to conan-center-index break our build again, how to avoid this?

### ISSUE THREAD

**Issue by cbecker** [conan-io/conan-center-index#3077]
Hi all. First, I'm not sure if this is the best place where to post this, I'll be happy to re-post wherever suits better the purpose.

Before I go into this, I want to say thanks for making and maintaining conan, it's a big relief for C/C++ programmers!

The issue I'm running into is that updates to conan-center-index have broken our build system repeatedly (I work at Invision AI), while we haven't touched any of the recipe versions we need. Our recent two examples are:

1. GDAL: we were using 3.10 and when 3.1.2 was introduced with https://github.com/conan-io/conan-center-index/pull/2478 it also back-ported changes to 3.10 and broke our build- we had to then hardcode the recipe version and modify our conanfile.py
2. xorg: similar issue, https://github.com/conan-io/conan-center-index/pull/2741 , but now the newest xorg recipe requires conan 1.29.0, and we were using 1.28.x. We didn't change anything in our conanfile.py, and if we delete the conan cache and rebuild from scratch, it doesn't work anymore. Solution: update conan to 1.29.0

My question is that I'm not sure what's the point of versioning if the recipes are not versioned as a whole, or I'm not understanding the philosophy behind versioning that is used by conan-center index. 

Given that dependencies are so crucial for a product, I'd expect that if a recipe is changed, that is reflected by modifying its version and that, by default, no 'automatic' updates are made. i.e., gdal 3.10 is fixed to what it was before 3.1.2 is introduced. Like that, there are no side-effects and old code can work accordingly.

**Comment by SSE4** [comment]
for 2., having the latest version of conan client is a mandatory requirements, as [documented](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md)
for 1., could you possibly be more specific, what exactly was broken in your build (e.g. error message or log would be helpful)?

**Comment by cbecker** [comment]
@SSE4 , thanks for the quick answer.

> for 2., having the latest version of conan client is a mandatory requirement, as [documented](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md)

This link you gave is about 'Adding Packages to ConanCenter'. Our problem comes when _consuming_ packages. Our setup that worked yesterday and the very same setup is broken today. No server is down. No file is unavailable. But an update (that we do not want) broke our build. Maybe the best option is to have our own version of `conan-center-index` and update only when necessary, and not use any recipe from `_/_`?

> for 1., could you possibly be more specific, what exactly was broken in your build (e.g. error message or log would be helpful)?
I don't remember exactly now, but it was an issue with qhull to do with components, which could link to conan not being at the latest version. Sadly I don't remember much more than that.

The main issue I see is that if we request in our `conanfile.py` a recipe, let's say, `GDAL/3.1.2`, we'd like to always get the same lib *and* recipe version. So, one can try to specify the revision for that library (GDAL), but GDAL may depend on other libs, and the latest will be pulled for those libs except if we specify in _each_ the revision we need. This becomes a burden to maintain, which is why we are concerned about how to keep our system stable and working regardless of the updates in conan-center index.

**Comment by SpaceIm** [comment]
I can't say for sure, but is it https://github.com/conan-io/conan-center-index/pull/2478 which broke your build system https://github.com/conan-io/conan-center-index/pull/2472 ? They were merged the same day.

Due to https://github.com/conan-io/conan-center-index/pull/2472, gdal recipe is broken until https://github.com/conan-io/conan-center-index/pull/2913 is merged.

**Comment by SSE4** [comment]
@cbecker yes, it's about adding packages to the conan center. if added package requires the latest version, then it requires the latest version for consumption as well

**Comment by SpaceIm** [comment]
@cbecker, you could require specific revision for each recipe in your dependency graph, not only direct dependencies.

**Comment by cbecker** [comment]
> @cbecker, you could require specific revision for each recipe in your dependency graph, not only direct dependencies.

Yes that's an option, but the respective conanfile.py would be hard to understand and pretty messy IMO.
I'd like to know how companies deal with this. It may be that the only way to go around it is to have all recipes in our own repos and don't use anything from upstream, which means that it's easy to diverge and also potentially less contribution to open source recipes. Does anyone know if this is what companies using conan do?

Personally it looks like something is off, as the way the updates are done to conan-center-index are likely to break previously-working code, and there is no easy mechanism to stick to a given version, except of hardcoding *every* dependency, which is an overkill IMO.

**Comment by prince-chrismc** [comment]
As a fellow enterprise consumer of CCI, I highly recommend following:

> you could require specific revision for each recipe in your dependency graph, not only direct dependencies.

> Does anyone know if this is what companies using conan do?

Yes, I believe it is.

You can find many example in the main Conan repo like this... https://github.com/conan-io/conan/issues/7574#issuecomment-678370963, where the Conan team has strongly encouraged this. It's a sticking point that has come up many times since I started to follow this effort.

> but the respective conanfile.py would be hard to understand and pretty messy IMO.

I would argue to this is true if you have very large monolithic projects... I know because I'm guilty of this... Newer smaller projects have proven advantageous for CI/CD pipelines and have fewer deps and are less painful in this respect.

> Just brainstorming: another option could be to have a sort of 'snapshot recipe revisions' feature in conan

Now, if you would like to attack this, https://docs.conan.io/en/latest/versioning/lockfiles.html seem to be what you are looking for... It's in my backlog but I would love to hear about your experience if you try it out!

**Comment by uilianries** [comment]
[Revisions](https://docs.conan.io/en/latest/versioning/revisions.html) and [lockfiles](https://docs.conan.io/en/latest/versioning/lockfiles.html) are the recommended features if you want a specific behavior. Otherwise, you will have to consume from the edge, which requires the latest Conan version available.

**Comment by cbecker** [comment]
Thanks for the pointers to revisions and lockfiles.

After discussing the options to make our builds stable, we decided to have our own clone of conan-center-index and avoid using the conan-center remote. With our setup this is not ideal but it is the method that will bring most stability while being very easy to version and be able to pinpoint versions.

I'll be looking forward to the progress with revisions and lockfiles nontheless. Thanks again for all the replies!

PS: I'm happy to close this issue, except if you want to keep it open for more comments.

**Comment by Croydon** [comment]
Conan got a lot of updates to revisions and lockfiles in the years since this issue was opened. I guess it can be closed now?
